File: web/src/shared/binding.js
/**
 * @typedef {Object} GitCommit
 * @property {string} hash
 * @property {string} parentHash
 * @property {string} name
 * @property {string} authorName
 * @property {string} authorDate
 */

/**
 * @typedef {Object} GitBranch
 * @property {string} name
 * @property {string} commitHash
 */

export const Binding = {
	// keep alphabetical order
	/** @type {() => Promise<string[]>} */
	getCurrentPath: window.getCurrentPath,
	/** @type {(path: string) => Promise<GitBranch[]>} */
	getGitBranches: window.getGitBranches,
	/** @type {(path: string, limit: number, offset: number) => Promise<GitCommit[]>} */
	getGitCommits: window.getGitCommits,
	/** @type {() => Promise<string>} */
	getGitVersion: window.getGitVersion,
	/** @type {() => Promise<string>} */
	getPathSeparator: window.getPathSeparator,
	/** @type {(path: string[]) => Promise<string[]>} */
	listDir: window.listDir,
	/** @type {(link: string) => Promise<void>} */
	openLink: window.openLink,
	/** @type {(path: string[]) => Promise<string>} */
	serializePath: window.serializePath,
}

File: internal/web_view/binding.go
package web_view

import (
	"webview-test/internal"
	"webview-test/internal/fs"
	"webview-test/internal/git"

	webview "github.com/webview/webview_go"
)

func BindWebView(webView webview.WebView, git git.Git) error {
	// keep alphabetical order
	if err := webView.Bind("getCurrentPath", fs.GetCurrentPath); err != nil {
		return err
	}
	if err := webView.Bind("getGitBranches", git.GetBranches); err != nil {
		return err
	}
	if err := webView.Bind("getGitCommits", git.GetCommits); err != nil {
		return err
	}
	if err := webView.Bind("getGitVersion", git.GetVersion); err != nil {
		return err
	}
	if err := webView.Bind("getPathSeparator", fs.GetPathSeparator); err != nil {
		return err
	}
	if err := webView.Bind("listDir", fs.ListDir); err != nil {
		return err
	}
	if err := webView.Bind("openLink", internal.OpenLink); err != nil {
		return err
	}
	if err := webView.Bind("serializePath", fs.SerializePath); err != nil {
		return err
	}
	return nil
}
